File: src/test/kotlin/com/djulia/transactions/InMemoryOnlyBankingTransactionServiceTest.java
package com.djulia.transactions;

import com.djulia.transactions.InMemoryOnlyBankingTransactionService.InactiveAccountException;
import com.djulia.transactions.InMemoryOnlyBankingTransactionService.InvalidWithdrawalAmountException;
import org.junit.Before;
import org.junit.Test;

import java.math.BigDecimal;

import static com.djulia.transactions.InMemoryOnlyBankingTransactionService.*;
import static com.djulia.transactions.InMemoryOnlyBankingTransactionService.InsufficientFundsException;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class InMemoryOnlyBankingTransactionServiceTest {
    private BankingTransactionService bankingTransactionService;
    private String accountNumberWith1500Balance = "123456789ABC";
    private String closedAccountNumber = "999912932";
    private String frozenAccountNumber = "723623921";

    @Before
    public void setup() {
        InMemoryAccountRepo accountRepo = new InMemoryAccountRepo();
        bankingTransactionService = new InMemoryOnlyBankingTransactionService(accountRepo);
        accountRepo.save(new Account(accountNumberWith1500Balance, new BigDecimal(1500), Account.Status.OPEN));
        accountRepo.save(new Account(closedAccountNumber, new BigDecimal(1500), Account.Status.CLOSED));
        accountRepo.save(new Account(frozenAccountNumber, new BigDecimal(1500), Account.Status.FROZEN));
    }

    @Test
    public void withdrawReturnsNewAccountWithLowerBalance_whenWithdrawalUnderCurrentBalance() throws Exception {
        Account result = bankingTransactionService.withdraw(accountNumberWith1500Balance, new BigDecimal(100));

        assertThat(result.getBalance()).isEqualTo(new BigDecimal(1400));
    }

    @Test
    public void withdrawThrowsError_whenAccountDoesNotExist() throws Exception {
        assertThatThrownBy(
                () -> bankingTransactionService.withdraw("not-a-real-account-number", new BigDecimal(100))
        ).isInstanceOf(NoSuchAccountException.class);

    }

    @Test
    public void withdrawingZeroDollarsOrLess_raisesAnError() throws Exception {
        assertThatThrownBy(
                () -> bankingTransactionService.withdraw(accountNumberWith1500Balance, new BigDecimal(0))
        ).isInstanceOf(InvalidWithdrawalAmountException.class);

        assertThatThrownBy(
                () -> bankingTransactionService.withdraw(accountNumberWith1500Balance, new BigDecimal(-1))
        ).isInstanceOf(InvalidWithdrawalAmountException.class);
    }

    @Test
    public void withdrawingMoreThanAccountBalance_raisesAnError() throws Exception {
        assertThatThrownBy(
                () -> bankingTransactionService.withdraw(accountNumberWith1500Balance, new BigDecimal(1501))
        ).isInstanceOf(InsufficientFundsException.class);
    }

    @Test
    public void withdrawingFromANonOpenAccount_raisesAnError() {
        assertThatThrownBy(
                () -> bankingTransactionService.withdraw(closedAccountNumber, new BigDecimal(20))
        ).isInstanceOf(InactiveAccountException.class);

        assertThatThrownBy(
                () -> bankingTransactionService.withdraw(frozenAccountNumber, new BigDecimal(20))
        ).isInstanceOf(InactiveAccountException.class);
    }

}

File: src/test/src/com/djulia/transactions/InMemoryOnlyBankingTransactionServiceTest.kt
package com.djulia.transactions

import org.junit.Before
import org.junit.Test

import java.math.BigDecimal
import org.assertj.core.api.KotlinAssertions.assertThat

class InMemoryOnlyBankingTransactionServiceTest {
    private lateinit var bankingTransactionService: BankingTransactionService
    private val accountNumberWith1500Balance = "123456789ABC"
    private val closedAccountNumber = "999912932"
    private val frozenAccountNumber = "723623921"

    @Before
    fun setup() {
        val accountRepo = InMemoryAccountRepo()
        bankingTransactionService = InMemoryOnlyBankingTransactionService(accountRepo)
        accountRepo.save(Account(accountNumberWith1500Balance, BigDecimal(1500), Account.Status.OPEN))
        accountRepo.save(Account(closedAccountNumber, BigDecimal(1500), Account.Status.CLOSED))
        accountRepo.save(Account(frozenAccountNumber, BigDecimal(1500), Account.Status.FROZEN))
    }


    @Test
    @Throws(Exception::class)
    fun withdrawReturnsNewAccountWithLowerBalance_whenWithdrawalUnderCurrentBalance() {
        val result = bankingTransactionService.withdraw(accountNumberWith1500Balance, BigDecimal(100))

        assertThat(result.updatedAccount?.balance).isEqualTo(BigDecimal(1400))
    }

    @Test
    @Throws(Exception::class)
    fun withdrawThrowsError_whenAccountDoesNotExist() {
        val result = bankingTransactionService.withdraw("not-a-real-account-number", BigDecimal(100))
        assertThat(result).isEqualTo(WithdrawalResult.error(WithdrawalResult.Error.NoSuchAccountError()))
    }

    @Test
    @Throws(Exception::class)
    fun withdrawingZeroDollarsOrLess_returnsAnError() {
        val withdrawZeroResult = bankingTransactionService.withdraw(accountNumberWith1500Balance, BigDecimal(0))
        assertThat(withdrawZeroResult).isEqualTo(WithdrawalResult.error(WithdrawalResult.Error.InvalidWithdrawalAmountError()))

        val negativeWithdrawalResult = bankingTransactionService.withdraw(accountNumberWith1500Balance, BigDecimal(-1))
        assertThat(negativeWithdrawalResult).isEqualTo(WithdrawalResult.error(WithdrawalResult.Error.InvalidWithdrawalAmountError()))
    }

    @Test
    @Throws(Exception::class)
    fun withdrawingMoreThanAccountBalance_returnsAnError() {
        val result = bankingTransactionService.withdraw(accountNumberWith1500Balance, BigDecimal(1501))
        assertThat(result).isEqualTo(WithdrawalResult.error(WithdrawalResult.Error.InsufficientFundsError(BigDecimal(1))))
    }

    @Test
    fun withdrawingFromANonOpenAccount_returnsAnError() {
        val closedAccountResult = bankingTransactionService.withdraw(closedAccountNumber, BigDecimal(20))
        assertThat(closedAccountResult).isEqualTo(WithdrawalResult.error(WithdrawalResult.Error.InactiveAccountError()))

        val frozenAccountResult = bankingTransactionService.withdraw(frozenAccountNumber, BigDecimal(20))
        assertThat(frozenAccountResult).isEqualTo(WithdrawalResult.error(WithdrawalResult.Error.InactiveAccountError()))
    }
}